### ISSUE TITLE
Synchronization failed together with Keepass Twofish plugin Version 1.4

### ISSUE THREAD

**Issue by gitmuti** [PhilippC/keepass2android#887]
There is a new Keepass Twofish cipher plugin version **1.4** (former was 1.3).
This Twofish plugin was adapted in order to work together with Keepass Version > 2.38

**Now the synchronization between Keepass2Android via Cloud to Keepass 2.42.1 with Twofish cipher plugin version 1.4 failed.**
**A detail description with steps** of the failure is described here:
[(https://sourceforge.net/p/keepass/discussion/329221/thread/0762d062df/?limit=25#8492](https://sourceforge.net/p/keepass/discussion/329221/thread/0762d062df/?limit=25#8492)

Perhaps it is a subsequent impact of the new Twofish plugin version 1.4.
The change made in this plugin version 1.4 is described here:
[https://sourceforge.net/p/keepass/discussion/329220/thread/e7599abb0a/](https://sourceforge.net/p/keepass/discussion/329220/thread/e7599abb0a/)

**Comment by crass** [comment]
Wow 2 years later... and still no progress. I guess not a lot of people are syncing Twofish encrypted databases between KP2A and official KeePass2. To be clear this makes Twofish encrypted databases **unusable** between newer versions of KeePass2 and Keepass2Android, can't even be opened. Also, there should be a way to write a database that can not be read according to the second link OP posted. I haven't tried this on KP2A, but I'd guess the bug is in there.

Unfortunately, its not enough to fix the padding issue because there's all the prior Twofish encrypted databases that have incorrect padding. For format 3.x, its not a big deal, as noted in that link. The Twofish plugin can write the correct padding and on read no check the padding, thus sending trailing garbage bytes up the next layer, which is the `HashedBlockStream`. Since the `HashedBlockStream` format has an end of stream (EOS) marker, everything after EOS will be ignored.

The problem comes with format 4.x, which has the CipherStream contained in an HmacBlockStream, where the CipherStream contains the raw XML or the XML compressed with gzip. If the database has no compression, the XML will have trailing garbage, on which the XML parser will throw an exception.

I propose that this be solved by upgrading the TwofishCipher plugin to 1.4, with [a patch](https://sourceforge.net/p/keepass/discussion/329221/thread/b8d76c5390/#90db/ffb9/59d5) that leaves the padding in the stream on read, but writes the correct padding (PKCS7). This then leaves the issue of the XML parser and trailing garbage for which I've created a [patch to the XML deserializer](https://sourceforge.net/p/keepass/patches/126/) so that it stops trying to parse xml when it reaches a KeePassFile end tag. The KeePassFile tag is the root element and so the end tag should indicated the end of the document and that anything after can be ignored.

These two patches fix this issue, while at the same time preserving compatibility for older Twofish encrypted databases with incorrect padding.

**Comment by PhilippC** [comment]
Indeed, I have never received any feedback except this issue about this. If you want to make this work, feel free to commit a patch in the repo here. Unfortunately I barely have time for updating the app these days.
